### PR TITLE
If server or UI port is unavailable, try to find an available one

### PR DESCRIPTION
**Pull request by lucasmls**
## What was changed

If the server or UI port is already in use by another process when running the CLI, the CLI will currently crash. The aim of this pull request is to modify this behavior; instead of crashing, the CLI will increment the port number by 1 until it finds an available port.

Screenshot 2023-10-27 at 21 53 04

---
I know this is a naive approach. I think a better way to handle this would be to ask the user to manually input an available port. I chose to stick with the easiest solution for the problem. However, if we all agree that prompting the user for the port is a better approach, I'm open to developing it.

**Another question**
Loren suggested that it would be nice if we added a log like this: _"Port 3000 is taken. Using port 3001 instead."_
I find it nice, but at the context that the code of this PR is being executed i don't have access to the logger. Any tips on this?

## Checklist

**1. Closes** 
https://github.com/temporalio/cli/issues/242

**2. How was this tested:**

- I've added the unit test `TestUIPortAlreadyInUse`, which locks the default ports and their subsequent ones (**server**: `7233` and `7234`, **ui**: `8233` and `8234`). By utilizing the previously coded helper `assertServerHealth`, I ensure that the server is running healthily.
- I also tested it manually. I opened another process using the default ports and ensured that the server was able to start and operate in the newly allocated ports.
Screenshot 2023-10-27 at 22 09 00

**3. Any docs updates needed?**

This alters some existing behavior. Should we include this new behavior in the CLI command description?

### CODE REVIEW

File: server/commands.go [hunk 1]
```diff
@@ -387,3 +388,30 @@ func getDynamicConfigValues(input []string) (map[dynamicconfig.Key][]dynamicconf
...
+		return false
+	}
+
+	if conn != nil {
+		_ = conn.Close()
+		return true
+	}
+
+	return false
+}
```
Comment (lucasmls) [inline server/commands.go:417]: I've added this code here for the sake of simplicity. I questioned whether it makes sense to move it to the [PortProvider](https://github.com/temporalio/cli/blob/main/server/config/freeport.go#L40) struct. Alternatively, if there's a better place for it, please let me know. What do all of you think?

**Comment by cretz** [comment]
>  The aim of this pull request is to modify this behavior; instead of crashing, the CLI will increment the port number by 1 until it finds an available port.

I am not sure it is desirable to use anything but the given/default port when starting the CLI (regardless of port purpose, UI, server, etc). There are programmatic expectations for ports when people start the CLI, we should not continue if we can't meet those expectations. Similarly, it that port is in use by another instance of the UI and someone starts the CLI, they will blindly go to the common UI port unaware it's another instance. We want users to solve the port ambiguity by picking another or stopping what is already using the port, not silently picking another for them.

If we want to suggest an alternative, like supporting `--ui-port random`, that may be more acceptable than changing existing behavior to instead swallow errors on fixed port ambiguity.

File: server/commands.go (hunk 1 above)
Comment (josh-berry) [reply server/commands.go:417]: I agree changing PortProvider, or just using it outright, makes sense. (I think it already has some functionality for picking a free port, so you don't have to reimplement this.)

**Review by josh-berry — COMMENTED** [review]
Thanks for sending this in!

I agree with Chad that this should be opt-in, so as to avoid breaking any scripts etc. that may have been written around the assumption that starting the server will fail if any of the ports are in use.

**Pull request closed without merging**